server/src-lib/Hasura/GraphQL/Schema/Remote.hs

```diff
@@ -221,20 +212,25 @@ remoteSchemaUnion
   -> m (Parser 'Output n ())
 remoteSchemaUnion schemaDoc defn@(G.UnionTypeDefinition description name _directives objectNames) =
   P.memoizeOn 'remoteSchemaObject defn do
-  objDefs <- traverse getObject objectNames
+  objDefs <- traverse (getObject schemaDoc name) objectNames
   objs :: [Parser 'Output n ()] <- traverse (remoteSchemaObject schemaDoc) objDefs
   when (null objs) $
     throw400 RemoteSchemaError $ "List of member types cannot be empty for union type " <> squote name
   pure $ void $ P.selectionSetUnion name description objs
-  where
-    getObject :: G.Name -> m G.ObjectTypeDefinition
-    getObject objectName =
-      onNothing (lookupObject schemaDoc objectName) $
-        case lookupInterface schemaDoc objectName of
-          Nothing -> throw400 RemoteSchemaError $ "Could not find type " <> squote objectName
-            <> ", which is defined as a member type of Union " <> squote name
-          Just _  -> throw400 RemoteSchemaError $ "Union type " <> squote name <>
-            " can only include object types. It cannot include " <> squote objectName
+
+getObject
+  :: MonadError QErr m
+  => SchemaIntrospection
+  -> G.Name
+  -> G.Name
+  -> m G.ObjectTypeDefinition
+getObject schemaDoc name objectName =
+  onNothing (lookupObject schemaDoc objectName) $
+    case lookupInterface schemaDoc objectName of
+      Nothing -> throw400 RemoteSchemaError $ "Could not find type " <> squote objectName
+        <> ", which is defined as a member type of " <> squote name
```

Comment:
```suggestion
        <> ", which is defined as a member type of Union " <> squote name
```

Reply:
I'm working on this. I think it might be slightly more involved. Will be done within 30min.